### PR TITLE
fix(buf_updates): send updates when putting past last line

### LINKED ISSUE #13710: treesitter: highlight is not updated when duplicating last line
- `nvim --version`:  0.5.0 (as of https://launchpad.net/~neovim-ppa/+archive/ubuntu/unstable)
- `vim -u DEFAULTS` (version: ) behaves differently?
- Operating system/version: GNU/Linux
- Terminal name/version: st 0.8.2
- `$TERM`: st-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC teste.c
```

Assuming nvim-treesitter is installed.

Run:

```vim
packadd nvim-treesitter
TSEnableAll highlight
```

Then write:

```c
#include <stdio.h>
```

Then use `yyp`.

### Actual behaviour
The new line is not highlighted.

### Expected behaviour
The new line should be highlighted.

### PR DESCRIPTION
**Pull request by vigoux**
Please @phelipetls, could you confirm this fixes your issue?

And I am not sure about the implementation, do you agree with the
changes @bfredl ?

Fixes #13710

### CODE REVIEW

File: src/nvim/ops.c [hunk 1]
```diff
@@ -3398,7 +3398,7 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
         bcount_t totsize = 0;
         int lastsize = 0;
         if (y_type == kMTCharWise
-            || (y_type == kMTLineWise && flags & PUT_LINE_SPLIT)) {
```
Comment (bfredl) [inline src/nvim/ops.c:3401]: hmm this looks messy and weird, and this if condition is not correct (need to duplicate the new check, not remove this one).

Can't the issue be solved just by adjusting the if-statement around `mark_adjust` below, which seems like the real root case? "Skip mark_adjust when adding lines after the last one"

Comment (bfredl) [reply src/nvim/ops.c:3401]: to be clear this is not just for buf_updates, there _can_ be an right-gravity ext-mark past the end of the buffer.

Comment (vigoux) [reply src/nvim/ops.c:3401]: The thing is that we lose the total put length in the mark_adjsut below, so that's my issue

Comment (bfredl) [reply src/nvim/ops.c:3401]: lose how? mark_adjust always calculates the length, this is not a special case.

Comment (vigoux) [reply src/nvim/ops.c:3401]: Oh you mean do the extmark move in mark_adjust ?

Comment (bfredl) [reply src/nvim/ops.c:3401]: Yes, just try loosen the condition, i see no reason it wouldn't work.

Comment (vigoux) [reply src/nvim/ops.c:3401]: eheh, it indeed works

Comment (vigoux) [reply src/nvim/ops.c]: I think I fixed the issue, could you confirm the fix suits you ?

Comment (bfredl) [reply src/nvim/ops.c]: Lgtm

**Comment by vigoux** [comment]
@bfredl any updates on this?

**Comment by bfredl** [comment]
I think i said LTGM but I can say it again :)

**Pull request merged**